### ISSUE TITLE
AbsentUfsPathCache is not respected on getStatus and listStatus calls

### ISSUE THREAD

**Issue by lilyzhoupeijie** [Alluxio/alluxio#12833]
**Alluxio Version:**
2.5.0-SNAPSHOT-9833d5f0841ef03fc9bb572fe4dccdd73bb98175

**Describe the bug**
when using spark write to alluxio,commit file takes a long time.
And we have alreay use the following config:
```
config("spark.hadoop.mapreduce.fileoutputcommitter.algorithm.version","2")
alluxio.user.file.writetype.default: "MUST_CACHE"
```
we use the following code to write to alluxio:
`data_df.write.mode("overwrite").format("tfrecord").option("recordType", "Example").save("alluxio://fluid1-master-0.kf-partition:20009/s3/test/2020-02-05")`
and there are logs:
```
2021-02-10 02:48:18,272 DEBUG UnderFileSystemWithLogging - Exit (Error): GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000977_5577, Error=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000977_5577
2021-02-10 02:48:18,272 WARN  UfsStatusCache - Failed to fetch status for /s3/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000977_5577: java.io.FileNotFoundException: s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000977_5577
2021-02-10 02:48:19,025 WARN  ObjectUnderFileSystem - Error fetching directory status, assuming directory s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000970_5570 does not exist
2021-02-10 02:48:19,026 DEBUG UnderFileSystemWithLogging - Exit (Error): GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000970_5570, Error=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000970_5570
2021-02-10 02:48:19,026 WARN  UfsStatusCache - Failed to fetch status for /s3/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000970_5570: java.io.FileNotFoundException: s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000970_5570
2021-02-10 02:48:19,351 WARN  ObjectUnderFileSystem - Error fetching directory status, assuming directory s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000958_5558 does not exist
2021-02-10 02:48:19,352 DEBUG UnderFileSystemWithLogging - Exit (Error): GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000958_5558, Error=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000958_5558
2021-02-10 02:48:19,352 WARN  UfsStatusCache - Failed to fetch status for /s3/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000958_5558: java.io.FileNotFoundException: s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000958_5558
2021-02-10 02:48:19,583 WARN  ObjectUnderFileSystem - Error fetching directory status, assuming directory s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000981_5581 does not exist
2021-02-10 02:48:19,583 DEBUG UnderFileSystemWithLogging - Exit (Error): GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000981_5581, Error=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000981_5581
2021-02-10 02:48:19,583 WARN  UfsStatusCache - Failed to fetch status for /s3/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000981_5581: java.io.FileNotFoundException: s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000981_5581
2021-02-10 02:48:20,478 WARN  ObjectUnderFileSystem - Error fetching directory status, assuming directory s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000965_5565 does not exist
2021-02-10 02:48:20,479 DEBUG UnderFileSystemWithLogging - Exit (Error): GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000965_5565, Error=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000965_5565
2021-02-10 02:48:20,479 WARN  UfsStatusCache - Failed to fetch status for /s3/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000965_5565: java.io.FileNotFoundException: s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000965_5565
2021-02-10 02:48:20,528 DEBUG UnderFileSystemWithLogging - Exit (OK): ListStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000983_5583
2021-02-10 02:48:20,528 DEBUG UnderFileSystemWithLogging - Enter: GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000983_5583
2021-02-10 02:48:20,576 DEBUG UnderFileSystemWithLogging - Exit (OK): ListStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000985_5585
2021-02-10 02:48:20,576 DEBUG UnderFileSystemWithLogging - Enter: GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000985_5585
2021-02-10 02:48:20,639 WARN  ObjectUnderFileSystem - Error fetching directory status, assuming directory s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000973_5573 does not exist
2021-02-10 02:48:20,639 DEBUG UnderFileSystemWithLogging - Exit (Error): GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000973_5573, Error=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000973_5573
2021-02-10 02:48:20,639 WARN  UfsStatusCache - Failed to fetch status for /s3/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000973_5573: java.io.FileNotFoundException: s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000973_5573
2021-02-10 02:48:20,674 DEBUG UnderFileSystemWithLogging - Exit (OK): ListStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000987_5587
2021-02-10 02:48:20,674 DEBUG UnderFileSystemWithLogging - Enter: GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000987_5587
2021-02-10 02:48:20,675 DEBUG UnderFileSystemWithLogging - Exit (OK): ListStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000991_5591
2021-02-10 02:48:20,675 DEBUG UnderFileSystemWithLogging - Enter: GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000991_5591
2021-02-10 02:48:20,889 DEBUG UnderFileSystemWithLogging - Exit (OK): ListStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000990_5590
2021-02-10 02:48:20,889 DEBUG UnderFileSystemWithLogging - Enter: GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000990_5590
2021-02-10 02:48:20,890 WARN  ObjectUnderFileSystem - Error fetching directory status, assuming directory s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000954_5554 does not exist
2021-02-10 02:48:20,890 DEBUG UnderFileSystemWithLogging - Exit (Error): GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000954_5554, Error=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000954_5554
2021-02-10 02:48:20,890 WARN  UfsStatusCache - Failed to fetch status for /s3/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000954_5554: java.io.FileNotFoundException: s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000954_5554
2021-02-10 02:48:20,894 DEBUG UnderFileSystemWithLogging - Exit (OK): ListStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000989_5589
2021-02-10 02:48:20,894 DEBUG UnderFileSystemWithLogging - Enter: GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000989_5589
2021-02-10 02:48:21,122 WARN  ObjectUnderFileSystem - Error fetching directory status, assuming directory s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000982_5582 does not exist
2021-02-10 02:48:21,123 DEBUG UnderFileSystemWithLogging - Exit (Error): GetStatus: path=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000982_5582, Error=s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000982_5582
2021-02-10 02:48:21,123 WARN  UfsStatusCache - Failed to fetch status for /s3/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000982_5582: java.io.FileNotFoundException: s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000982_5582
2021-02-10 02:48:21,701 DEBUG UnderFileSystemWithLogging - Enter: GetSpace: path=/underFSStorage, type=SPACE_USED
2021-02-10 02:48:21,701 DEBUG UnderFileSystemWithLogging - Exit (OK): GetSpace: path=/underFSStorage, type=SPACE_USED
2021-02-10 02:48:21,702 DEBUG UnderFileSystemWithLogging - Enter: GetSpace: path=/underFSStorage, type=SPACE_FREE
2021-02-10 02:48:21,702 DEBUG UnderFileSystemWithLogging - Exit (OK): GetSpace: path=/underFSStorage, type=SPACE_FREE
2021-02-10 02:48:21,702 DEBUG UnderFileSystemWithLogging - Enter: GetSpace: path=/underFSStorage, type=SPACE_TOTAL
2021-02-10 02:48:21,702 DEBUG UnderFileSystemWithLogging - Exit (OK): GetSpace: path=/underFSStorage, type=SPACE_TOTAL
2021-02-10 02:48:22,171 WARN  ObjectUnderFileSystem - Error fetching directory status, assuming directory s3://fluid/test/2020-02-05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000971_5571 does not exist
```
**Urgency**
Writing  efficiency is greatly affected and this  blocks the launch of alluxio in our company

Attached master logs for a new run:

[fluid1-master-0-alluxio-master.log](https://github.com/Alluxio/alluxio/files/6065014/fluid1-master-0-alluxio-master.log)

**Comment by apc999** [comment]
@lilyzhoupeijie @cheyang  the reason we see a lot of messages like "GetStatus: path=s3://fluid/test/2020-02--05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000977_5577" even when `alluxio.user.file.writetype.default=MUST_CACHE` is because the default metadata load type is `ONCE`. So Spark tasks  in your case is about to write a temp file to `s3://fluid/test/2020-02--05/_temporary/0/_temporary/attempt_20210210023417_0005_m_000977_5577`, however, on Spark side it first checks existence of this file (that's why you see `GetStatus` calls and this is outside of Alluxio). 

There are two solutions:
1. Set `alluxio.user.file.metadata.load.type=NEVER` with `alluxio.user.file.writetype.default=MUST_CACHE`. In this case, the Spark job will never load metadata from UFS. The potential downside is that if you also need to read something from UFS, it will be tricky.
2. Let's add some logic to avoid loading metadata from certain blacklist, like the persistence blacklist. In this case, if we see the path to check existence is `*_temporary*`, we will only check Alluxio space but never loading from UFS.

Which option works better for you?

**Comment by apc999** [comment]
I am investigating the absent cache in Alluxio which should help your case---if it works as expected

**Comment by apc999** [comment]
After investigation, I believe this is a regression that the `AbsentUfsPathCache` is not respected on getStatus and listStatus calls. Note that for metadata operations (e.g., committing files for Spark/MR/Hive jobs), this can be a huge overhead.

To be specific, the AbsentUfsPathCache is not leveraged when the temp root directory for output is fetched from UFS. 
One can find it from the following subset of error messages. In here each Failed attempt to read from UFS takes 4sec, adding a lot of execution time to commit final results for Spark job

```
2021-02-10 02:34:24,905 DEBUG UnderFileSystemWithLogging - Enter: ListStatus: path=s3://fluid/test/2020-02-05
2021-02-10 02:34:27,895 DEBUG UnderFileSystemWithLogging - Exit (OK): ListStatus: path=s3://fluid/test/2020-02-05
2021-02-10 02:34:27,895 DEBUG UnderFileSystemWithLogging - Enter: GetStatus: path=s3://fluid/test/2020-02-05
2021-02-10 02:34:30,893 DEBUG UnderFileSystemWithLogging - Exit (Error): GetStatus: path=s3://fluid/test/2020-02-05, Error=s3://fluid/test/2020-02-05
2021-02-10 02:34:30,893 WARN  UfsStatusCache - Failed to fetch status for /s3/test/2020-02-05: java.io.FileNotFoundException: s3://fluid/test/2020-02-05
2021-02-10 02:34:30,894 DEBUG UnderFileSystemWithLogging - Enter: ListStatus: path=s3://fluid/test/2020-02-05/_temporary
2021-02-10 02:34:33,788 DEBUG UnderFileSystemWithLogging - Exit (OK): ListStatus: path=s3://fluid/test/2020-02-05/_temporary
2021-02-10 02:34:33,788 DEBUG UnderFileSystemWithLogging - Enter: GetStatus: path=s3://fluid/test/2020-02-05/_temporary
2021-02-10 02:34:37,221 WARN  ObjectUnderFileSystem - Error fetching directory status, assuming directory s3://fluid/test/2020-02-05/_temporary does not exist
2021-02-10 02:34:37,221 DEBUG UnderFileSystemWithLogging - Exit (Error): GetStatus: path=s3://fluid/test/2020-02-05/_temporary, Error=s3://fluid/test/2020-02-05/_temporary
```

**Comment by yuzhu** [comment]
@lilyzhoupeijie @cheyang  Taking a deeper look into the issue here. What is your value for the configuration `alluxio.user.file.metadata.sync.interval`?

**Comment by yuzhu** [comment]
Some update after I took a closer look. 
@apc999 you are correct, the absent cache is currently not used in sync situations. I am not sure it should be. When the semantics of the sync interval asks for a sync to be performed, we have to sync. It does not matter that we saw this file not existing on the UFS before. 

Currently, the only use of the absent cache is when sync is not needed and loadMetadata is ONCE (the default). In that case, the absent cache saves us a trip to the UFS because we know this path does not exist at some point in the past (we looked at it ONCE already), so according to the ONCE semantics, so we can return a FileNotFound exception right away. 

Syncing behavior is governed by a different cache UfsSyncPathCache. In that case, if a file or directory has been synced in the last sync interval will be cached and we will know not to sync. 

In this particular case, there are two reasons for the repeated reads to the UFS, and it does not have anything to do with sync. (It was actually triggered by LoadMetadata) 
1. Absent cache is asynchronous so the cache content did not get updated right away, causing a repeated read into the UFS.
2. To make 1 worse, our UfsStatusCache failed to find the file, and our ufsSyncStream did not find it in the UfsStatusCache, so it went to the UFS to retrieve it a second time. This is what you saw in the log. 

I will make a fix for the second issue and make absent cache synchronous because it probably is faster to insert an entry to the cache than visiting the UFS.
